core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -106,7 +107,66 @@
         /**
          * Unassigned as a result of a failed primary while the replica was initializing.
          */
-        PRIMARY_FAILED;
+        PRIMARY_FAILED
+    }
+
+    /**
+     * Captures the status of an unsuccessful allocation attempt for the shard,
+     * causing it to remain in the unassigned state.
+     *
+     * Note, ordering of the enum is important, make sure to add new values
+     * at the end and handle version serialization properly.
+     */
+    public enum AllocationAttempt implements Writeable {
```

Comment:
also make all the methods  referring to this use this name (like `getAllocationStatus`)? I think it's clearer..